plugins/fapolicyd.c

```diff
@@ -0,0 +1,175 @@
+{
+    int fd = -1;
+    struct stat s;
+
+    fd = open(state->fifo_path, O_WRONLY);
+    if(fd == -1) {
+        rpmlog(RPMLOG_DEBUG, "Open: %s -> %s\n", state->fifo_path, strerror(errno));
+        return RPMRC_OK;
+    }
+
+    if (stat(state->fifo_path, &s) == -1) {
+        rpmlog(RPMLOG_DEBUG, "Stat: %s -> %s\n", state->fifo_path, strerror(errno));
+        close(fd);
+        return RPMRC_OK;
+    } else {
+
+        if (!S_ISFIFO(s.st_mode)) {
+            rpmlog(RPMLOG_DEBUG, "File: %s exists but it is not a pipe!\n", state->fifo_path);
+            close(fd);
+            return RPMRC_OK;
+        }
+
+        // we require pipe to have 0660 permissions
+        if (!(
+            (s.st_mode & S_IRUSR) &&
+            (s.st_mode & S_IWUSR) &&
+            !(s.st_mode & S_IXUSR) &&
+
+            (s.st_mode & S_IRGRP) &&
+            (s.st_mode & S_IWGRP) &&
+            !(s.st_mode & S_IXGRP) &&
+
+            !(s.st_mode & S_IROTH) &&
+            !(s.st_mode & S_IWOTH) &&
+            !(s.st_mode & S_IXOTH) )) {
+
+            rpmlog(RPMLOG_ERR, "File: %s has 0%d%d%d instead of 0660 \n",
+                    state->fifo_path,
+                    ((s.st_mode & S_IRUSR) ? 4 : 0) +
+                    ((s.st_mode & S_IWUSR) ? 2 : 0) +
+                    ((s.st_mode & S_IXUSR) ? 1 : 0),
+
+                    ((s.st_mode & S_IRGRP) ? 4 : 0) +
+                    ((s.st_mode & S_IWGRP) ? 2 : 0) +
+                    ((s.st_mode & S_IXGRP) ? 1 : 0),
+
+                    ((s.st_mode & S_IROTH) ? 4 : 0) +
+                    ((s.st_mode & S_IWOTH) ? 2 : 0) +
+                    ((s.st_mode & S_IXOTH) ? 1 : 0) );
+
+            close(fd);
+            return RPMRC_OK;
+        }
+    }
+
+    state->fd = fd;
+    state->ready = 1;
+    return RPMRC_OK;
+}
+
+static rpmRC write_fifo(struct fapolicyd_data* state, const char * str, size_t len)
+{
+    ssize_t ret = write(state->fd, str, len);
+
+    if (ret == -1) {
+        rpmlog(RPMLOG_DEBUG, "Write: %s -> %s\n", state->fifo_path, strerror(errno));
+    }
+
+    return RPMRC_OK;
+}
+
+static rpmRC fapolicyd_init(rpmPlugin plugin, rpmts ts)
+{
+    if (rpmtsFlags(ts) & (RPMTRANS_FLAG_TEST))
+        return RPMRC_OK;
+
+    open_fifo(&fapolicyd_state);
+    return RPMRC_OK;
+}
+
+static void fapolicyd_cleanup(rpmPlugin plugin)
+{
+    if (fapolicyd_state.ready)
+        close(fapolicyd_state.fd);
+
+    fapolicyd_state.ready = 0;
+    fapolicyd_state.fd = -1;
+}
+
+static rpmRC fapolicyd_tsm_post(rpmPlugin plugin, rpmts ts, int res)
+{
+    if (fapolicyd_state.ready)
+        write_fifo(&fapolicyd_state, "1", 2);
+
+    return RPMRC_OK;
+}
+
+static rpmRC fapolicyd_scriptlet_pre(rpmPlugin plugin, const char *s_name,
+				    int type)
+{
+    if (fapolicyd_state.ready && fapolicyd_state.first_scriptlet) {
+        // send signal to flush cache
+        write_fifo(&fapolicyd_state, "2", 2);
+        fapolicyd_state.first_scriptlet = 0;
```

Comment:
Hmm, doesn't this need to take place once per package, rather than once per transaction? (I could be misunderstanding and/or misremembering the discussion though)

Reply:
We need a hook to be placed right after all files are installed on the disk and before the execution of the first scriptlet.
After flushing of the object cache we can handle execution of the scriptlets with integrity checking on.

Reply:
Hum, I sense a misunderstanding here. There can be no such global hook, as rpm interleaves file operations and scriptlets. Even on per-package basis there can first be various pre-scriptlets, then the files get laid down, and then there can be more scriptlets running. And note that for the pre-scriptlet stuff you do *not* want cache flushes as no files have been changed yet, so tracking the "first scriptlet" is not what you want to do at all.

AIUI, what you really need to be doing is set a global "files changed" flag on every entry to fsm_file_prepare, and then from scriptlet_pre flush the cache *and* reset the changed variable for the next package. With that it'll track the real situation and eg packages without files will not cause unnecessary cache flushes.

Of course it's also possible I'm misunderstanding something here...

Reply:
I rewrote the plugin with the "files changed" part and I resolved most of the issues.